rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,261 @@
+- Pipelines have access to any components from the root configuration. If the transform `foo` is defined in `/etc/vector/bar.toml`, it will be accessible by the pipeline `/etc/vector/pipelines/pipeline.toml`.
+- If no pipeline is defined, Vector behaves as if the feature didn't exist. This way, a configuration from a version without the `pipeline` feature will keep working.
+- If a pipeline file is left empty, Vector behaves as if it doesn't exist.
+
+If any of the following constraints are not valid, Vector will error on boot. If this occurs during a reload, an error will be triggered and handled in the same fashion as other reload errors.
+
+- There cannot be several pipelines with the same id (for example `load-balancer.yml` and `load-balancer.json`).
+- The pipeline's configuration files should only contain transforms.
+- A pipeline's transform cannot have the same name as any component from the root configuration.
+- A pipeline cannot use another pipeline's component.
+
+### Representation
+
+As mentioned in the previous section, a pipeline is _just_ a set of transforms.
+
+To be able to forward the events going through the pipeline to a `sink`, we need add a new concept of `route`.
+
+`route` components are used solely to build the topology and represent an interface between the topology and the external sinks.
+
+A pipeline will have the following internal representation before building the topology.
+
+```rust
+struct Route {
+  inputs: Vec<String>,
+  outputs: Vec<String>,
+}
+struct PipelineConfigBuilder {
+  id: String,
+  transforms: Map<String, TransformOuter>,
+  routes: Map<String, Route>,
+}
+```
+
+Which will create the following configuration file.
+
+```toml
+# /etc/vector/pipelines/pipeline.toml
+[transforms.foo]
+type = "remap"
+inputs = ["from-root"]
+# ...
+
+[transforms.bar]
+type = "remap"
+inputs = ["foo"]
+# ...
+
+[routes.hot-storage]
+inputs = ["foo"]
+outputs = ["dc1", "dc2"]
+
+[routes.cold-storage]
+inputs = ["bar"]
+outputs = ["dc-us", "dc-eu"]
+```
+
+The `Route` structure is made to forward the events from inside the pipeline to an external component.
+
+### From configuration to topology
+
+If we look deeper at the configuration building process, the configuration compiler will require the pipelines to build the [configuration](https://github.com/timberio/vector/blob/v0.15.0/src/config/builder.rs#L71).
+
+To do so, we'll need to implement a `PipelineConfigBuilder` from the previous section. We'll then update [the `compile` function](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L4) to build a `Config` containing the required pipeline components. The compiler will load the pipeline's transforms and add the route `outputs` to the corresponding `sinks`.
+
+The components coming from the pipeline will be cloned inside the final `Config`, in the `transforms` `IndexMap` and the `Routes` from the pipeline components will be added to the referring components input field.
```

Comment:
Can't tell if this is just missing a comma between `transforms` and `IndexMap` or if the latter three code highlights are unrelated. Either way, I'm having a hard time parsing it.

Reply:
> In the `IndexMap` containing the `tranforms`